### PR TITLE
Make use of relaxed*() methods in PoolThreadCache.MemoryRegionCache when possible

### PR DESCRIPTION
**Pull request by laosijikaichele**
Motivation:

This is inspired by https://github.com/netty/netty/pull/11960.
We should be able to make use of `MessagePassingQueue.relaxedPoll()` methods in `PoolThreadCache.MemoryRegionCache` to reduce the overhead.

Modification:

Use `MessagePassingQueue.relaxedPoll(...)`.

Result:

Less overhead.

### CODE REVIEW

File: buffer/src/main/java/io/netty/buffer/PoolThreadCache.java [hunk 1]
```diff
@@ -368,7 +369,7 @@ protected abstract void initBuf(PoolChunk<T> chunk, ByteBuffer nioBuffer, long h
         @SuppressWarnings("unchecked")
         public final boolean add(PoolChunk<T> chunk, ByteBuffer nioBuffer, long handle, int normCapacity) {
             Entry<T> entry = newEntry(chunk, nioBuffer, handle, normCapacity);
-            boolean queued = queue.offer(entry);
+            boolean queued = queue.relaxedOffer(entry);
```
Comment (franz1981) [inline buffer/src/main/java/io/netty/buffer/PoolThreadCache.java:372]: relaxed offer on mpsc doesn't make much difference https://github.com/JCTools/JCTools/blob/master/jctools-core/src/main/java/org/jctools/queues/MpscArrayQueue.java#L456

while `relaxedPoll` would save it to spin awaiting the elements (and breaking linearizability if the queue);

I don't suggest to go there unless there's some evidence of pathologic cases which can by fixing by this change...

Comment (laosijikaichele) [reply buffer/src/main/java/io/netty/buffer/PoolThreadCache.java:372]: > relaxed offer on mpsc doesn't make much difference

True for `relaxedOffer`.

About the `relaxedPoll` part, I may try some benchmark if possible.

Comment (normanmaurer) [reply buffer/src/main/java/io/netty/buffer/PoolThreadCache.java]: @laosijikaichele any update ?

Comment (normanmaurer) [reply buffer/src/main/java/io/netty/buffer/PoolThreadCache.java]: please ping me once you are ready

**Comment by normanmaurer** [comment]
Let me close this for now.. Please reopen if you think this is still worthwhile

**Comment by laosijikaichele** [comment]
@normanmaurer @franz1981  Sorry for let this hang... I made a benchmark, and it seems there is no obvious difference between `relaxedPoll` and `poll` in this scenario. So it's ok to close this.

The following is the benchmark code, in case anyone is interested:
```
@Warmup(iterations = 3)
@Measurement(iterations = 3)
public class MemoryRegionCacheBenchmark extends AbstractMicrobenchmark {
    private final PooledByteBufAllocator allocator = new PooledByteBufAllocator();
    private final ByteBuf[] byteBufs = new ByteBuf[259];

    @Setup
    public void setup() throws InterruptedException {
        // Allocate some byteBuf.
        for (int i = 0 ; i < byteBufs.length; i++) {
            byteBufs[i] = allocator.heapBuffer();
        }
        // Release these byteBuf to cache in another thread.
        Thread t0 = new Thread(new Runnable() {
            @Override
            public void run() {
                for (int i = 0 ; i < byteBufs.length; i++) {
                    byteBufs[i].release();
                }
            }
        });
        t0.start();
        t0.join();
    }

    @Benchmark
    @BenchmarkMode(Mode.Throughput)
    public void allocateFromCache(Blackhole blackhole) {
        // Allocate some byteBuf from cache.
        for (int i = 0 ; i < byteBufs.length; i++) {
            byteBufs[i] = allocator.heapBuffer();
            blackhole.consume(byteBufs[i]);
        }
    }

    @TearDown(Level.Invocation)
    public void release() throws InterruptedException {
        // Release these byteBuf to cache in another thread.
        Thread t1 = new Thread(new Runnable() {
            @Override
            public void run() {
                for (int i = 0 ; i < byteBufs.length; i++) {
                    byteBufs[i].release();
                }
            }
        });
        t1.start();
        t1.join();
    }
}
```

**Pull request closed without merging**